### ISSUE TITLE
Clipping Planes cause visual artifacts and don't work properly at all times

### ISSUE THREAD

**Issue by Gmod4phun** [Facepunch/sbox-issues#2530]
### Describe the bug

SceneObjects that have clip planes set and clipping enabled do not behave correctly at all times.  From what I can tell, there are multiple issues with them.

1. Opaque objects have black artifacts in the part where the object should not be rendered: this does not happen with transparent objects, or when lowering the alpha of an opaque object (e.q. a wooden box has black artifacts, transparent wineglass model does not). I don't know why this is caused, I can only guess it has something to do with the depth buffer.

2. The clipping plane starts/stops working when looking in a certain direction: this pretty much causes the objects to appear clipped and then not clipped, depending on what angle they are viewed from (possibly more factors, I don't know).

3. Objects seem to use incorrect clipping planes at times: having multiple objects, with each having a specific clipping plane, they seem to sometimes ignore it and instead use the plane of another object. I also had a crash when stopping the game after playing around with bunch of these, not sure if there might be a memory leak or something.

This pretty much makes the clipping planes highly unreliable at the moment.

I included a few videos showcasing the issues.

### To Reproduce

1. Spawn the following test entity/multiple entities:

```
using Sandbox;

[Spawnable]
public partial class TestBox : Prop
{
	public override void Spawn()
	{
		base.Spawn();
		SetModel( "models/citizen_props/crate01.vmdl" );
		SetupPhysicsFromModel( PhysicsMotionType.Static );
	}

	[Event.Frame]
	public void Frame()
	{
		var plane = new Plane( WorldSpaceBounds.Center, Rotation.Up );

		SceneObject.Attributes.SetCombo( "D_ENABLE_USER_CLIP_PLANE", true );
		SceneObject.Attributes.Set( "ClipPlane0", new Vector4( plane.Normal, plane.Distance ) );
	}
}
```

2. Play around with them and see the issues.

### Expected behavior

Expected behaviour is no visual artifacts and proper functioning of the clipping planes.

### Media/Files

https://user-images.githubusercontent.com/5991682/201152744-7a5f7e24-c221-463e-87b9-6292736bb59b.mp4

https://user-images.githubusercontent.com/5991682/201152767-72268559-05bd-404f-84ad-277a0546b482.mp4

https://user-images.githubusercontent.com/5991682/201152792-59eed800-b656-4f80-890c-63e03b4426a0.mp4

**Comment by MuffinTastic** [comment]
From things I've observed before I think there's two issues at play here.

One seems to be something to do with lower resolution shadow maps
![image](https://user-images.githubusercontent.com/10884425/201324474-82cd5340-0430-449e-b3ca-fd4972b328c3.png)

The other I believe might have something to do with SceneObject batching.  Of a given batched set of SceneObjects within the player's view, the engine is apparently only using the attributes of one of them, even if the others within that batch have those attributes set differently.  Others have run into this issue in mildly different situations as well:  #1604, #2105

**Comment by Gmod4phun** [comment]
Bump. This is even more broken now.

**Comment by PixelQubed** [comment]
Yeah, this is a problem. Big problem. What if you want to clip weapons from going through walls like CSGO or make portals that do their job?
